Chrash at moving structure

```
[m[33m[15:21:28] [Render thread/WARN] [minecraft/ClientPlayNetHandler]: Unknown custom packet identifier: spawnermod:main
[m[32m[15:21:28] [Render thread/INFO] [Waila/]: Received config from the server: {"minecraft:hide_infestations":true}
[m[32m[15:21:28] [Render thread/INFO] [minecraft/AdvancementList]: Loaded 511 advancements
[m[1;31m[15:22:13] [Server thread/ERROR] [minecraft/PointOfInterestData]: POI data mismatch: never registered at BlockPos{x=231, y=128, z=-130}
[m[1;31m[15:22:13] [Server thread/ERROR] [minecraft/MinecraftServer]: Encountered an unexpected exception
net.minecraft.crash.ReportedException: Ticking entity
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:853) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:784) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:114) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:637) [?:?] {re:classloading,pl:accesstransformer:B}
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_262] {}
Caused by: java.lang.NullPointerException
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionCollider.isCollidingWithWorld(ContraptionCollider.java:260) ~[?:mc1.15.2_v0.2.4e+custom] {re:classloading}
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionCollider.collideBlocks(ContraptionCollider.java:215) ~[?:mc1.15.2_v0.2.4e+custom] {re:classloading}
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionEntity.func_70071_h_(ContraptionEntity.java:161) ~[?:mc1.15.2_v0.2.4e+custom] {re:classloading}
	at net.minecraft.world.server.ServerWorld.func_217479_a(ServerWorld.java:600) ~[?:?] {re:classloading}
	at net.minecraft.world.World.func_217390_a(World.java:536) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.world.server.ServerWorld.func_72835_b(ServerWorld.java:401) ~[?:?] {re:classloading}
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:849) ~[?:?] {re:classloading,pl:accesstransformer:B}
	... 4 more
[m[1;31m[15:22:13] [Server thread/ERROR] [minecraft/MinecraftServer]: This crash report has been saved to: /home/jan/.local/share/multimc/instances/create --/.minecraft/crash-reports/crash-2020-10-19_15.22.13-server.txt
[m[32m[15:22:13] [Server thread/INFO] [minecraft/MinecraftServer]: Stopping server
[m[32m[15:22:13] [Server thread/INFO] [minecraft/MinecraftServer]: Saving players
[m[32m[15:22:13] [Server thread/INFO] [minecraft/ServerPlayNetHandler]: Hero9909 lost connection: Server closed
[m[32m[15:22:13] [Server thread/INFO] [minecraft/MinecraftServer]: Hero9909 left the game
[m[1;31m[15:22:13] [Server thread/ERROR] [minecraft/MinecraftServer]: Exception stopping the server
java.lang.IllegalStateException: Removing entity while ticking!
	at net.minecraft.world.server.ServerWorld.removeEntity(ServerWorld.java:1001) ~[?:?] {re:classloading}
	at net.minecraft.world.server.ServerWorld.removePlayer(ServerWorld.java:1022) ~[?:?] {re:classloading}
	at net.minecraft.world.server.ServerWorld.func_217434_e(ServerWorld.java:1018) ~[?:?] {re:classloading}
	at net.minecraft.server.management.PlayerList.func_72367_e(PlayerList.java:315) ~[?:?] {re:classloading}
	at net.minecraft.network.play.ServerPlayNetHandler.func_147231_a(ServerPlayNetHandler.java:909) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.network.NetworkManager.func_179293_l(NetworkManager.java:365) ~[?:?] {re:classloading}
	at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213167_f(SourceFile:79) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.network.play.ServerPlayNetHandler.func_194028_b(ServerPlayNetHandler.java:272) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.management.PlayerList.func_72392_r(PlayerList.java:708) ~[?:?] {re:classloading}
	at net.minecraft.server.MinecraftServer.func_71260_j(MinecraftServer.java:557) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.integrated.IntegratedServer.func_71260_j(IntegratedServer.java:216) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:676) [?:?] {re:classloading,pl:accesstransformer:B}
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_262] {}
[m[32m[15:22:13] [Render thread/INFO] [STDOUT/]: [net.minecraft.util.registry.Bootstrap:func_179870_a:103]: ---- Minecraft Crash Report ----
// Don't be sad, have a hug! <3

Time: 10/19/20 3:22 PM
Description: Ticking entity

java.lang.NullPointerException: Ticking entity
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionCollider.isCollidingWithWorld(ContraptionCollider.java:260) ~[?:mc1.15.2_v0.2.4e+custom] {re:classloading}
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionCollider.collideBlocks(ContraptionCollider.java:215) ~[?:mc1.15.2_v0.2.4e+custom] {re:classloading}
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionEntity.func_70071_h_(ContraptionEntity.java:161) ~[?:mc1.15.2_v0.2.4e+custom] {re:classloading}
	at net.minecraft.world.server.ServerWorld.func_217479_a(ServerWorld.java:600) ~[?:?] {re:classloading}
	at net.minecraft.world.World.func_217390_a(World.java:536) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.world.server.ServerWorld.func_72835_b(ServerWorld.java:401) ~[?:?] {re:classloading}
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:849) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:784) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:114) ~[?:?] {re:classloading,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:637) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_262] {}


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Render thread
Stacktrace:
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionCollider.isCollidingWithWorld(ContraptionCollider.java:260)
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionCollider.collideBlocks(ContraptionCollider.java:215)
	at com.simibubi.create.content.contraptions.components.structureMovement.ContraptionEntity.func_70071_h_(ContraptionEntity.java:161)
	at net.minecraft.world.server.ServerWorld.func_217479_a(ServerWorld.java:600)

-- Entity being ticked --
Details:
	Entity Type: create:stationary_contraption (com.simibubi.create.content.contraptions.components.structureMovement.ContraptionEntity)
	Entity ID: 1435
	Entity Name: Stationary Contraption
	Entity's Exact location: 225.00, 136.00, -136.01
	Entity's Block location: World: (225,136,-137), Chunk: (at 1,8,7 in 14,-9; contains blocks 224,0,-144 to 239,255,-129), Region: (0,-1; contains chunks 0,-32 to 31,-1, blocks 0,0,-512 to 511,255,-1)
	Entity's Momentum: -0.00, -0.00, -0.01
	Entity's Passengers: []
	Entity's Vehicle: ~~ERROR~~ NullPointerException: null
Stacktrace:
	at net.minecraft.world.World.func_217390_a(World.java:536)
	at net.minecraft.world.server.ServerWorld.func_72835_b(ServerWorld.java:401)

-- Affected level --
Details:
	All players: 1 total; [ServerPlayerEntity['Hero9909'/1039, l='Kelgra', x=230.06, y=135.00, z=-134.45]]
	Chunk stats: ServerChunkCache: 3050
	Level dimension: DimensionType{minecraft:overworld}
	Level name: Kelgra
	Level seed: 8467000433611163313
	Level generator: ID 02 - largeBiomes, ver 0. Features enabled: true
	Level generator options: {}
	Level spawn location: World: (1,63,0), Chunk: (at 1,3,0 in 0,0; contains blocks 0,0,0 to 15,255,15), Region: (0,0; contains chunks 0,0 to 31,31, blocks 0,0,0 to 511,255,511)
	Level time: 1391051 game time, 1472771 day time
	Known server brands: forge
	Level was modded: true
	Level storage version: 0x04ABD - Anvil
	Level weather: Rain time: 98111 (now: false), thunder time: 81332 (now: false)
	Level game mode: Game mode: survival (ID 0). Hardcore: false. Cheats: false
Stacktrace:
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:849)
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:784)
	at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:114)
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:637)
	at java.lang.Thread.run(Thread.java:748)

-- System Details --
Details:
	Minecraft Version: 1.15.2
	Minecraft Version ID: 1.15.2
	Operating System: Linux (amd64) version 5.4.0-48-generic
	Java Version: 1.8.0_262, Azul Systems, Inc.
	Java VM Version: OpenJDK 64-Bit Server VM (mixed mode), Azul Systems, Inc.
	Memory: 1054791584 bytes (1005 MB) / 1901592576 bytes (1813 MB) up to 3817865216 bytes (3641 MB)
	CPUs: 16
	JVM Flags: 2 total; -Xms512m -Xmx4096m
	ModLauncher: 5.1.0+69+master.79f13f7
	ModLauncher launch target: fmlclient
	ModLauncher naming: srg
	ModLauncher services: 
		/eventbus-2.2.0-service.jar eventbus PLUGINSERVICE 
		/forge-1.15.2-31.2.19-launcher.jar object_holder_definalize PLUGINSERVICE 
		/forge-1.15.2-31.2.19-launcher.jar runtime_enum_extender PLUGINSERVICE 
		/accesstransformers-2.1.1-shadowed.jar accesstransformer PLUGINSERVICE 
		/forge-1.15.2-31.2.19-launcher.jar capability_inject_definalize PLUGINSERVICE 
		/forge-1.15.2-31.2.19-launcher.jar runtimedistcleaner PLUGINSERVICE 
		/forge-1.15.2-31.2.19-launcher.jar fml TRANSFORMATIONSERVICE 
	FML: 31.2
	Forge: net.minecraftforge:31.2.19
	FML Language Providers: 
		javafml@31.2
		minecraft@1
	Mod List: 
		forge-1.15.2-31.2.19-client.jar Minecraft {minecraft@1.15.2 DONE}
		branders-enhanced-spawnermod-1.15.2-1.6.5.jar Enhanced Mob Spawners {spawnermod@1.6.5 DONE}
		Bookshelf-1.15.2-5.6.40.jar Bookshelf {bookshelf@5.6.40 DONE}
		EnchantmentDescriptions-1.15.2-2.0.9.jar EnchantmentDescriptions {enchdesc@2.0.9 DONE}
		forge-1.15.2-31.2.19-universal.jar Forge {forge@31.2.19 DONE}
		create-mc1.15.2_v0.2.4e.jar Create {create@mc1.15.2_v0.2.4e+custom DONE}
		Apotheosis-1.15.2-3.3.7.jar Apotheosis {apotheosis@3.3.7 DONE}
		ConfigurableCane-1.15.2-1.5.2.jar Configurable Cane {configurablecane@1.5.2 DONE}
		Placebo-1.15.2-3.1.0.jar Placebo {placebo@3.1.0 DONE}
		Hwyla-forge-1.10.8-B72_1.15.2.jar Waila {waila@version DONE}
		jei-1.15.2-6.0.3.15.jar Just Enough Items {jei@6.0.3.15 DONE}
		anviltweaks-2.3.jar Anvil Tweaks {anviltweaks@2.3 DONE}
	Player Count: 1 / 8; [ServerPlayerEntity['Hero9909'/1039, l='Kelgra', x=230.06, y=135.00, z=-134.45]]
	Data Packs: vanilla, mod:spawnermod (incompatible), mod:bookshelf (incompatible), mod:enchdesc (incompatible), mod:forge (incompatible), mod:create, mod:apotheosis (incompatible), mod:configurablecane (incompatible), mod:placebo (incompatible), mod:waila (incompatible), mod:jei (incompatible), mod:anviltweaks (incompatible)
	Type: Integrated Server (map_client.txt)
	Is Modded: Definitely; Client brand changed to 'forge'
```

Mod version v0.2.4e
minecraft 1.15.2

this chrash happens consistently if i try to move a structure, anything you would need to figure this out?

--- a/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/AbstractContraptionEntity.java
+++ b/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/AbstractContraptionEntity.java
@@ -435,6 +435,15 @@ public abstract class AbstractContraptionEntity extends Entity implements IEntit
 		}
 	}
 
+	@SuppressWarnings("deprecation")
+	@Override
+	public void remove(boolean keepData) {
+		if (!world.isRemote && !removed && contraption != null) {
+			contraption.stop(world);
+		}
+		super.remove(keepData);
+	}
+
 	protected abstract StructureTransform makeStructureTransform();
 
 	@Override

--- a/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/Contraption.java
+++ b/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/Contraption.java
@@ -743,7 +743,6 @@ public abstract class Contraption {
 	}
 
 	public void addBlocksToWorld(World world, StructureTransform transform) {
-		stop(world);
 		for (boolean nonBrittles : Iterate.trueAndFalse) {
 			for (BlockInfo block : blocks.values()) {
 				if (nonBrittles == BlockMovementTraits.isBrittle(block.state))

--- a/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/TranslatingContraption.java
+++ b/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/TranslatingContraption.java
@@ -1,5 +1,6 @@
 package com.simibubi.create.content.contraptions.components.structureMovement;
 
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
 
@@ -16,7 +17,7 @@ public abstract class TranslatingContraption extends Contraption {
 
 	public Set<BlockPos> getColliders(World world, Direction movementDirection) {
 		if (getBlocks() == null)
-			return null;
+			return Collections.EMPTY_SET;
 		if (cachedColliders == null || cachedColliderDirection != movementDirection) {
 			cachedColliders = new HashSet<>();
 			cachedColliderDirection = movementDirection;

--- a/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/piston/LinearActuatorTileEntity.java
+++ b/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/piston/LinearActuatorTileEntity.java
@@ -78,8 +78,10 @@ public abstract class LinearActuatorTileEntity extends KineticTileEntity impleme
 				else
 					sendData();
 				return;
+			} else {
+				if (getSpeed() != 0)
+					assemble();
 			}
-			assemble();
 			return;
 		}
 

--- a/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/piston/PistonContraption.java
+++ b/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/piston/PistonContraption.java
@@ -53,9 +53,14 @@ public class PistonContraption extends TranslatingContraption {
 	public boolean assemble(World world, BlockPos pos) {
 		if (!collectExtensions(world, pos, orientation))
 			return false;
+		int count = blocks.size();
 		if (!searchMovedStructure(world, anchor, retract ? orientation.getOpposite() : orientation))
 			return false;
-		bounds = bounds.union(pistonExtensionCollisionBox);
+		if (blocks.size() == count) { // no new blocks added
+			bounds = pistonExtensionCollisionBox;
+		} else {
+			bounds = bounds.union(pistonExtensionCollisionBox);
+		}
 		startMoving(world);
 		return true;
 	}
@@ -112,8 +117,10 @@ public class PistonContraption extends TranslatingContraption {
 		anchor = pos.offset(direction, initialExtensionProgress + 1);
 		extensionLength = extensionsInBack + extensionsInFront;
 		initialExtensionProgress = extensionsInFront;
-		pistonExtensionCollisionBox = new AxisAlignedBB(end.offset(direction, -extensionsInFront)
-			.subtract(anchor));
+		pistonExtensionCollisionBox = new AxisAlignedBB(
+				BlockPos.ZERO.offset(direction, -1),
+				BlockPos.ZERO.offset(direction, -extensionLength - 1)).expand(1,
+						1, 1);
 
 		if (extensionLength == 0)
 			return false;
@@ -124,7 +131,7 @@ public class PistonContraption extends TranslatingContraption {
 			BlockPos relPos = pole.pos.offset(direction, -extensionsInFront);
 			BlockPos localPos = relPos.subtract(anchor);
 			getBlocks().put(localPos, new BlockInfo(localPos, pole.state, null));
-			pistonExtensionCollisionBox = pistonExtensionCollisionBox.union(new AxisAlignedBB(localPos));
+			//pistonExtensionCollisionBox = pistonExtensionCollisionBox.union(new AxisAlignedBB(localPos));
 		}
 
 		return true;

--- a/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/train/capability/MinecartController.java
+++ b/src/main/java/com/simibubi/create/content/contraptions/components/structureMovement/train/capability/MinecartController.java
@@ -19,10 +19,16 @@ import com.simibubi.create.foundation.utility.Couple;
 import com.simibubi.create.foundation.utility.Iterate;
 import com.simibubi.create.foundation.utility.VecHelper;
 
+import net.minecraft.block.BlockState;
+import net.minecraft.block.PoweredRailBlock;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.item.minecart.AbstractMinecartEntity;
+import net.minecraft.entity.item.minecart.MinecartEntity;
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.NBTUtil;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.Constants.NBT;
@@ -78,8 +84,45 @@ public class MinecartController implements INBTSerializable<CompoundNBT> {
 				internalStall.booleanValue() || otherCart == null || !otherCart.isPresent() || otherCart.isStalled(false));
 
 		}));
-		if (!world.isRemote)
+		if (!world.isRemote) {
 			setStalled(internalStall.booleanValue(), true);
+			disassemble(cart);
+		}
+	}
+
+	private void disassemble(AbstractMinecartEntity cart) {
+		if (cart instanceof MinecartEntity) {
+			return;
+		}
+		List<Entity> passengers = cart.getPassengers();
+		if (passengers.isEmpty() || !(passengers.get(0) instanceof AbstractContraptionEntity)) {
+			return;
+		}
+		World world = cart.world;
+		int i = MathHelper.floor(cart.getX());
+		int j = MathHelper.floor(cart.getY());
+		int k = MathHelper.floor(cart.getZ());
+		if (world.getBlockState(new BlockPos(i, j - 1, k))
+				.isIn(BlockTags.RAILS)) {
+			--j;
+		}
+		BlockPos blockpos = new BlockPos(i, j, k);
+		BlockState blockstate = world.getBlockState(blockpos);
+		if (cart.canUseRail() && blockstate.isIn(BlockTags.RAILS)
+				&& blockstate.getBlock() instanceof PoweredRailBlock
+				&& ((PoweredRailBlock) blockstate.getBlock())
+						.isActivatorRail()) {
+			if (cart.isBeingRidden()) {
+				cart.removePassengers();
+			}
+
+			if (cart.getRollingAmplitude() == 0) {
+				cart.setRollingDirection(-cart.getRollingDirection());
+				cart.setRollingAmplitude(10);
+				cart.setDamage(50.0F);
+				cart.velocityChanged = true;
+			}
+		}
 	}
 
 	public boolean isFullyCoupled() {

